[OTEL Infra Host Mapping ]datadog.host.use_as_metadata causes Panic

**Problem** When enabling the OTEL Infra Host Mapping functionality the OTEL Collector is thrown into a panic that originates from this module. 

**Config:**
```
processors:
  transform/datadog:
    metric_statements: &statements
      - context: resource
        statements:
          - set(attributes["datadog.host.use_as_metadata"], true)
          - set(attributes["datadog.host.name"], <name>)
    log_statements: *statements
    trace_statements: *statements

exporters:
  datadog:
    hostname: <also try setting here>
    api: <ommiting>

```

**Full Trace**
```
panic: assignment to entry in nil map

goroutine 181 [running]:
github.com/DataDog/opentelemetry-mapping-go/pkg/inframetadata/internal/hostmap.(*HostMap).Update(0xc002e96e20, {0xc00252c570, 0x24}, {0xc003ff14b8?, 0xc003fa163c?})
	github.com/DataDog/opentelemetry-mapping-go/pkg/inframetadata@v0.13.4/internal/hostmap/hostmap.go:243 +0x619
github.com/DataDog/opentelemetry-mapping-go/pkg/inframetadata.(*Reporter).ConsumeResource(0xc002eb5290, {0xc003ff14b8?, 0xc003fa163c?})
	github.com/DataDog/opentelemetry-mapping-go/pkg/inframetadata@v0.13.4/reporter.go:137 +0x125
github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.consumeResource(0xc003b1d870?, {0xc003ff14b8, 0xc003fa163c}, 0x407985?)
	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter@v0.96.0/factory.go:73 +0x37
github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.(*metricsExporter).PushMetricsData(0xc002e8d560, {0x9216420, 0xc0022a32c0}, {0xc003fb3a10?, 0xc003fa163c?})
	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter@v0.96.0/metrics_exporter.go:203 +0x12d3
github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.(*metricsExporter).PushMetricsDataScrubbed(0xc002e8d560, {0x9216420?, 0xc0022a32c0?}, {0xc003fb3a10?, 0xc003fa163c?})
	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter@v0.96.0/metrics_exporter.go:185 +0x2c
go.opentelemetry.io/collector/exporter/exporterhelper.(*metricsRequest).Export(0x0?, {0x9216420?, 0xc0022a32c0?})
	go.opentelemetry.io/collector/exporter@v0.96.0/exporterhelper/metrics.go:59 +0x31
go.opentelemetry.io/collector/exporter/exporterhelper.(*timeoutSender).send(0xc00274ec78?, {0x9216420?, 0xc0022a32c0?}, {0x91cd7f0?, 0xc002e7cca8?})
	go.opentelemetry.io/collector/exporter@v0.96.0/exporterhelper/timeout_sender.go:43 +0x48
go.opentelemetry.io/collector/exporter/exporterhelper.(*baseRequestSender).send(0xc002e86380?, {0x9216420?, 0xc0022a32c0?}, {0x91cd7f0?, 0xc002e7cca8?})
	go.opentelemetry.io/collector/exporter@v0.96.0/exporterhelper/common.go:35 +0x30
go.opentelemetry.io/collector/exporter/exporterhelper.(*metricsSenderWithObservability).send(0xc002eb5da0, {0x92168d0?, 0xc001d7df50?}, {0x91cd7f0?, 0xc002e7cca8?})
	go.opentelemetry.io/collector/exporter@v0.96.0/exporterhelper/metrics.go:155 +0x7e
go.opentelemetry.io/collector/exporter/exporterhelper.newQueueSender.func1({0x92168d0?, 0xc001d7df50?}, {0x91cd7f0?, 0xc002e7cca8?})
	go.opentelemetry.io/collector/exporter@v0.96.0/exporterhelper/queue_sender.go:95 +0x84
go.opentelemetry.io/collector/exporter/internal/queue.(*boundedMemoryQueue[...]).Consume(0x9221bc0, 0xc002eb5e00)
	go.opentelemetry.io/collector/exporter@v0.96.0/internal/queue/bounded_memory_queue.go:57 +0xc7
go.opentelemetry.io/collector/exporter/internal/queue.(*Consumers[...]).Start.func1()
	go.opentelemetry.io/collector/exporter@v0.96.0/internal/queue/consumers.go:43 +0x79
created by go.opentelemetry.io/collector/exporter/internal/queue.(*Consumers[...]).Start in goroutine 1
	go.opentelemetry.io/collector/exporter@v0.96.0/internal/queue/consumers.go:39 +0x7d
panic: assignment to entry in nil map

goroutine 124 [running]:
github.com/DataDog/opentelemetry-mapping-go/pkg/inframetadata/internal/hostmap.(*HostMap).Update(0xc002e96e20, {0xc00252c570, 0x24}, {0xc003fb59f8?, 0xc002253790?})
	github.com/DataDog/opentelemetry-mapping-go/pkg/inframetadata@v0.13.4/internal/hostmap/hostmap.go:243 +0x619
github.com/DataDog/opentelemetry-mapping-go/pkg/inframetadata.(*Reporter).ConsumeResource(0xc002eb5290, {0xc003fb59f8?, 0xc002253790?})
	github.com/DataDog/opentelemetry-mapping-go/pkg/inframetadata@v0.13.4/reporter.go:137 +0x125
github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.consumeResource(0xc0034677f8?, {0xc003fb59f8, 0xc002253790}, 0xc0034679b8?)
	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter@v0.96.0/factory.go:73 +0x37
github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.(*metricsExporter).PushMetricsData(0xc002e8d560, {0x9216420, 0xc002cf2f60}, {0xc002e7cb70?, 0xc002253790?})
	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter@v0.96.0/metrics_exporter.go:203 +0x12d3
github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter.(*metricsExporter).PushMetricsDataScrubbed(0xc002e8d560, {0x9216420?, 0xc002cf2f60?}, {0xc002e7cb70?, 0xc002253790?})
	github.com/open-telemetry/opentelemetry-collector-contrib/exporter/datadogexporter@v0.96.0/metrics_exporter.go:185 +0x2c
go.opentelemetry.io/collector/exporter/exporterhelper.(*metricsRequest).Export(0x0?, {0x9216420?, 0xc002cf2f60?})
	go.opentelemetry.io/collector/exporter@v0.96.0/exporterhelper/metrics.go:59 +0x31
go.opentelemetry.io/collector/exporter/exporterhelper.(*timeoutSender).send(0xc002a76390?, {0x9216420?, 0xc002cf2f60?}, {0x91cd7f0?, 0xc002e7cc00?})
	go.opentelemetry.io/collector/exporter@v0.96.0/exporterhelper/timeout_sender.go:43 +0x48
go.opentelemetry.io/collector/exporter/exporterhelper.(*baseRequestSender).send(0xc002e86380?, {0x9216420?, 0xc002cf2f60?}, {0x91cd7f0?, 0xc002e7cc00?})
	go.opentelemetry.io/collector/exporter@v0.96.0/exporterhelper/common.go:35 +0x30
go.opentelemetry.io/collector/exporter/exporterhelper.(*metricsSenderWithObservability).send(0xc002eb5da0, {0x92168d0?, 0xc001d7dd00?}, {0x91cd7f0?, 0xc002e7cc00?})
	go.opentelemetry.io/collector/exporter@v0.96.0/exporterhelper/metrics.go:155 +0x7e
go.opentelemetry.io/collector/exporter/exporterhelper.newQueueSender.func1({0x92168d0?, 0xc001d7dd00?}, {0x91cd7f0?, 0xc002e7cc00?})
	go.opentelemetry.io/collector/exporter@v0.96.0/exporterhelper/queue_sender.go:95 +0x84
go.opentelemetry.io/collector/exporter/internal/queue.(*boundedMemoryQueue[...]).Consume(0x9221bc0, 0xc002eb5e00)
	go.opentelemetry.io/collector/exporter@v0.96.0/internal/queue/bounded_memory_queue.go:57 +0xc7
go.opentelemetry.io/collector/exporter/internal/queue.(*Consumers[...]).Start.func1()
	go.opentelemetry.io/collector/exporter@v0.96.0/internal/queue/consumers.go:43 +0x79
created by go.opentelemetry.io/collector/exporter/internal/queue.(*Consumers[...]).Start in goroutine 1
	go.opentelemetry.io/collector/exporter@v0.96.0/internal/queue/consumers.go:39 +0x7d
```


**Investigation:** It looks like this is originating from the ConsumeResource call on the Reporter struct. This struct appear to have an uninitialized hostmap

--- a/.chloggen/mx-psi_more-resilient-host-metadata.yaml
+++ b/.chloggen/mx-psi_more-resilient-host-metadata.yaml
@@ -1,0 +1,16 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component (e.g. pkg/quantile)
+component: pkg/inframetadata
+
+# A brief description of the change. Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Initialize gohai fields on first access to avoid nil pointer dereference
+
+# The PR related to this change
+issues: [306]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:

--- a/pkg/inframetadata/gohai/gohai.go
+++ b/pkg/inframetadata/gohai/gohai.go
@@ -82,17 +82,29 @@ type Payload struct {
 }
 
 // Platform returns a reference to the Gohai payload 'platform' map.
+// It initializes the field if nil.
 func (p *Payload) Platform() map[string]string {
+	if p.Gohai.Gohai.Platform == nil {
+		p.Gohai.Gohai.Platform = map[string]string{}
+	}
 	return p.Gohai.Gohai.Platform
 }
 
 // CPU returns a reference to the Gohai payload 'cpu' map.
+// It initializes the field if nil.
 func (p *Payload) CPU() map[string]string {
+	if p.Gohai.Gohai.CPU == nil {
+		p.Gohai.Gohai.CPU = map[string]string{}
+	}
 	return p.Gohai.Gohai.CPU
 }
 
 // Network returns a reference to the Gohai payload 'network' map.
+// It initializes the field if nil.
 func (p *Payload) Network() map[string]any {
+	if p.Gohai.Gohai.Network == nil {
+		p.Gohai.Gohai.Network = map[string]any{}
+	}
 	return p.Gohai.Gohai.Network
 }
 

--- a/pkg/inframetadata/payload/payload.go
+++ b/pkg/inframetadata/payload/payload.go
@@ -77,3 +77,15 @@ type Meta struct {
 	// HostAliases are other available host names
 	HostAliases []string `json:"host_aliases,omitempty"`
 }
+
+// NewEmpty creates a new HostMetadata with empty fields.
+// Pointer fields are initialized to empty structs.
+// All other fields are initialized to their zero value.
+func NewEmpty() HostMetadata {
+	return HostMetadata{
+		Meta:      &Meta{},
+		Tags:      &HostTags{},
+		Payload:   gohai.NewEmpty(),
+		Processes: &gohai.ProcessesPayload{},
+	}
+}

